File: src/end_to_end/sequence_planner.py
"""
Sequence Planner for end-to-end project creation.

Plans sequences and shots from story structure and dialogue, including:
- Shot distribution logic
- Camera angle/movement selection
- Dialogue timing integration
- Prompt module generation
"""

import uuid
from typing import List, Dict
from src.end_to_end.data_models import (
    ParsedPrompt, StoryStructure, DialogueScript, WorldConfig,
    SequencePlan, Sequence, Shot, PromptModules
)


class SequencePlanner:
    """Plans sequences and shots from story and dialogue"""
    
    def __init__(self):
        """Initialize sequence planner"""
        self.camera_angles = self._init_camera_angles()
        self.camera_movements = self._init_camera_movements()
        self.shot_types = self._init_shot_types()
    
    def _init_camera_angles(self) -> List[str]:
        """Initialize camera angle options"""
        return [
            "eye level",
            "low angle",
            "high angle",
            "dutch angle",
            "over the shoulder",
            "bird's eye view",
            "worm's eye view"
        ]
    
    def _init_camera_movements(self) -> List[str]:
        """Initialize camera movement options"""
        return [
            "static",
            "pan left",
            "pan right",
            "tilt up",
            "tilt down",
            "dolly in",
            "dolly out",
            "tracking shot",
            "crane shot"
        ]
    
    def _init_shot_types(self) -> Dict[str, str]:
        """Initialize shot type descriptions"""
        return {
            "establishing": "wide shot establishing the location",
            "close-up": "close-up focusing on character emotion",
            "medium": "medium shot showing character interaction",
            "wide": "wide shot showing full scene",
            "extreme close-up": "extreme close-up on specific detail",
            "two-shot": "two-shot framing two characters"
        }
    
    def plan_sequences(
        self,
        parsed_prompt: ParsedPrompt,
        story_structure: StoryStructure,
        dialogue_script: DialogueScript,
        world_config: WorldConfig
    ) -> SequencePlan:
        """
        Plan complete sequence and shot breakdown
        
        Args:
            parsed_prompt: Parsed user prompt
            story_structure: Story structure
            dialogue_script: Dialogue script
            world_config: World configuration
            
        Returns:
            Complete SequencePlan object
        """
        sequence_id = str(uuid.uuid4())
        
        # Generate sequences from story acts
        sequences = self._generate_sequences(
            story_structure,
            dialogue_script,
            parsed_prompt,
            world_config
        )
        
        # Calculate totals
        total_shots = sum(len(seq.shots) for seq in sequences)
        
        return SequencePlan(
            sequence_id=sequence_id,
            total_duration=parsed_prompt.duration_seconds,
            sequences=sequences,
            total_shots=total_shots
        )
    
    def _generate_sequences(
        self,
        story_structure: StoryStructure,
        dialogue_script: DialogueScript,
        parsed_prompt: ParsedPrompt,
        world_config: WorldConfig
    ) -> List[Sequence]:
        """Generate sequences from story acts"""
        sequences = []
        
        for act in story_structure.acts:
            sequence = self._generate_sequence(
                act,
                dialogue_script,
                parsed_prompt,
                world_config
            )
            sequences.append(sequence)
        
        return sequences
    
    def _generate_sequence(
        self,
        act,
        dialogue_script: DialogueScript,
        parsed_prompt: ParsedPrompt,
        world_config: WorldConfig
    ) -> Sequence:
        """Generate a single sequence from an act"""
        sequence_id = str(uuid.uuid4())
        
        # Determine shot count for this sequence
        shot_count = self._calculate_shot_count(
            act.duration,
            parsed_prompt.video_type
        )
        
        # Generate shots
        shots = self._generate_shots(
            shot_count,
            act,
            dialogue_script,
            parsed_prompt,
            world_config
        )
        
        # Determine mood and visual direction
        mood = self._determine_sequence_mood(act, parsed_prompt)
        visual_direction = self._determine_visual_direction(act, world_config)
        
        return Sequence(
            sequence_id=sequence_id,
            name=act.name,
            duration=act.duration,
            shots=shots,
            mood=mood,
            visual_direction=visual_direction
        )
    
    def _calculate_shot_count(self, duration: int, video_type: str) -> int:
        """Calculate number of shots for sequence duration"""
        # Shot duration varies by video type
        shot_durations = {
            "trailer": 3,  # Fast cuts
            "teaser": 4,
            "short_film": 6,  # Longer shots
            "music_video": 2,  # Very fast cuts
            "commercial": 5,
            "default": 5
        }
        
        avg_shot_duration = shot_durations.get(
            video_type.lower(),
            shot_durations["default"]
        )
        
        # Calculate shot count
        shot_count = max(2, duration // avg_shot_duration)
        return shot_count
    
    def _generate_shots(
        self,
        shot_count: int,
        act,
        dialogue_script: DialogueScript,
        parsed_prompt: ParsedPrompt,
        world_config: WorldConfig
    ) -> List[Shot]:
        """Generate shots for a sequence"""
        shots = []
        shot_duration = act.duration // shot_count
        
        # Get dialogue lines for this act's scenes
        act_dialogue_lines = self._get_act_dialogue_lines(act, dialogue_script)
        
        for i in range(shot_count):
            shot = self._generate_shot(
                i + 1,
                shot_duration,
                act,
                parsed_prompt,
                world_config,
                act_dialogue_lines,
                shot_count
            )
            shots.append(shot)
        
        return shots
    
    def _get_act_dialogue_lines(self, act, dialogue_script: DialogueScript) -> List:
        """Get dialogue lines for an act's scenes"""
        lines = []
        for scene in dialogue_script.scenes:
            if scene.scene_id in act.scenes:
                lines.extend(scene.dialogue_lines)
        return lines
    
    def _generate_shot(
        self,
        shot_number: int,
        duration: int,
        act,
        parsed_prompt: ParsedPrompt,
        world_config: WorldConfig,
        dialogue_lines: List,
        total_shots: int
    ) -> Shot:
        """Generate a single shot"""
        shot_id = str(uuid.uuid4())
        
        # Determine shot type based on position
        shot_type = self._determine_shot_type(shot_number, total_shots)
        
        # Select camera angle and movement
        camera_angle = self._select_camera_angle(shot_number, shot_type)
        camera_movement = self._select_camera_movement(shot_number, shot_type)
        
        # Determine lighting and composition
        lighting = self._determine_lighting(world_config, act.act_number)
        composition = self._determine_composition(shot_type)
        
        # Generate description
        description = self._generate_shot_description(
            shot_type,
            act,
            parsed_prompt,
            dialogue_lines,
            shot_number
        )
        
        # Generate prompt modules
        prompt_modules = self._generate_prompt_modules(
            parsed_prompt,
            world_config,
            shot_type,
            camera_angle,
            lighting
        )
        
        return Shot(
            shot_id=shot_id,
            shot_number=shot_number,
            duration=duration,
            description=description,
            camera_angle=camera_angle,
            camera_movement=camera_movement,
            lighting=lighting,
            composition=composition,
            prompt_modules=prompt_modules
        )
    
    def _determine_shot_type(self, shot_number: int, total_shots: int) -> str:
        """Determine shot type based on position in sequence"""
        if shot_number == 1:
            return "establishing"
        elif shot_number == total_shots:
            return "wide"
        elif shot_number % 3 == 0:
            return "close-up"
        elif shot_number % 2 == 0:
            return "medium"
        else:
            return "two-shot"
    
    def _select_camera_angle(self, shot_number: int, shot_type: str) -> str:
        """Select camera angle for shot"""
        if shot_type == "establishing":
            return "high angle"
        elif shot_type == "close-up":
            return "eye level"
        elif shot_type == "wide":
            return "eye level"
        else:
            # Cycle through angles
            return self.camera_angles[shot_number % len(self.camera_angles)]
    
    def _select_camera_movement(self, shot_number: int, shot_type: str) -> str:
        """Select camera movement for shot"""
        if shot_type == "establishing":
            return "dolly in"
        elif shot_type == "close-up":
            return "static"
        elif shot_type == "wide":
            return "pan right"
        else:
            # Cycle through movements
            return self.camera_movements[shot_number % len(self.camera_movements)]
    
    def _determine_lighting(self, world_config: WorldConfig, act_number: int) -> str:
        """Determine lighting for shot"""
        base_lighting = world_config.lighting_style
        
        # Adjust based on act
        if act_number == 1:
            return f"{base_lighting}, establishing mood"
        elif act_number == 3:
            return f"{base_lighting}, dramatic intensity"
        else:
            return f"{base_lighting}, building tension"
    
    def _determine_composition(self, shot_type: str) -> str:
        """Determine composition for shot"""
        compositions = {
            "establishing": "rule of thirds, balanced frame",
            "close-up": "centered subject, shallow depth of field",
            "medium": "balanced composition, natural framing",
            "wide": "expansive frame, environmental context",
            "extreme close-up": "tight frame, detail focus",
            "two-shot": "balanced two-person frame"
        }
        return compositions.get(shot_type, "balanced composition")
    
    def _generate_shot_description(
        self,
        shot_type: str,
        act,
        parsed_prompt: ParsedPrompt,
        dialogue_lines: List,
        shot_number: int
    ) -> str:
        """Generate shot description"""
        base_desc = self.shot_types.get(shot_type, "shot of the scene")
        
        # Add context
        desc = f"{base_desc} in {parsed_prompt.setting}"
        
        # Add act context
        if act.act_number == 1:
            desc += ", introducing the world"
        elif act.act_number == 3:
            desc += ", showing the resolution"
        else:
            desc += ", developing the conflict"
        
        return desc
    
    def _generate_prompt_modules(
        self,
        parsed_prompt: ParsedPrompt,
        world_config: WorldConfig,
        shot_type: str,
        camera_angle: str,
        lighting: str
    ) -> PromptModules:
        """Generate prompt modules for shot"""
        # Base prompt
        base = f"{shot_type} of {parsed_prompt.setting}"
        
        # Style prompt
        style_elements = ", ".join(parsed_prompt.visual_style[:3])
        style = f"{parsed_prompt.genre} style, {style_elements}"
        
        # Lighting prompt
        lighting_prompt = f"{lighting}, {world_config.atmosphere} atmosphere"
        
        # Composition prompt
        composition = f"{camera_angle} angle, cinematic framing"
        
        # Camera prompt
        camera = f"professional cinematography, {shot_type} composition"
        
        return PromptModules(
            base=base,
            style=style,
            lighting=lighting_prompt,
            composition=composition,
            camera=camera
        )
    
    def _determine_sequence_mood(self, act, parsed_prompt: ParsedPrompt) -> str:
        """Determine mood for sequence"""
        if act.act_number == 1:
            return parsed_prompt.mood[0] if parsed_prompt.mood else "mysterious"
        elif act.act_number == 3:
            return "intense"
        else:
            return "tense"
    
    def _determine_visual_direction(self, act, world_config: WorldConfig) -> str:
        """Determine visual direction for sequence"""
        return f"{world_config.visual_style[0]} aesthetic with {world_config.lighting_style}"

File: src/llm/sequence-planner.ts
/**
 * StoryCore Sequence Planner
 * Generates shot sequences with timing, duration, camera movements, and first image prompts
 */

import { AspectRatio } from './prompt-parser';

export interface SequenceShot {
  // Identification
  shotId: string;
  sequenceNumber: number;
  shotNumber: number;
  
  // Timing
  startTime: number;
  endTime: number;
  duration: number;
  
  // Scene Context
  sceneDescription: string;
  sceneIndex: number;
  
  // Camera
  shotType: ShotType;
  cameraAngle: CameraAngle;
  cameraMovement: CameraMovement;
  lens: LensType;
  
  // Visual
  firstImagePrompt: string;
  negativePrompt: string;
  styleNotes: string[];
  
  // Technical
  aspectRatio: AspectRatio;
  resolution: string;
  
  // Transitions
  transitionIn: TransitionType;
  transitionOut: TransitionType;
  
  // Audio
  musicDescription: string;
  soundEffects: string[];
  dialogue: DialogueLine[];
  
  // Metadata
  mood: string;
  intensity: 'low' | 'medium' | 'high';
}

export interface DialogueLine {
  speaker: string;
  text: string;
  startTime: number;
  endTime: number;
}

export type ShotType = 'ELS' | 'LS' | 'FS' | 'MCU' | 'CU' | 'ECU';
export type CameraAngle = 'eye-level' | 'low-angle' | 'high-angle' | 'bird-eye' | 'worm-eye';
export type CameraMovement = 'static' | 'dolly-in' | 'dolly-out' | 'pan-left' | 'pan-right' | 'tilt-up' | 'tilt-down' | 'tracking' | 'steadicam';
export type LensType = 'wide' | 'normal' | 'telephoto' | 'fisheye';
export type TransitionType = 'cut' | 'fade' | 'dissolve' | 'wipe' | 'match-cut';

export interface SequencePlan {
  planId: string;
  projectTitle: string;
  createdAt: string;
  
  // Configuration
  aspectRatio: AspectRatio;
  totalDuration: number;
  totalShots: number;
  
  // Sequences
  sequences: SequenceShot[];
  
  // Summary
  shotTypeDistribution: Record<ShotType, number>;
  musicMood: string;
  overallMood: string;
}

export interface SequencePlannerConfig {
  aspectRatio?: AspectRatio;
  targetDuration?: number;
  shotsPerMinute?: number;
  includeAudio?: boolean;
  includeDialogue?: boolean;
}

const DEFAULT_CONFIG: SequencePlannerConfig = {
  aspectRatio: '16:9',
  targetDuration: 60,
  shotsPerMinute: 3,
  includeAudio: true,
  includeDialogue: true,
};

export class SequencePlanner {
  private config: SequencePlannerConfig;

  constructor(config: Partial<SequencePlannerConfig> = {}) {
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  /**
   * Generate complete sequence plan from parsed prompt
   */
  generateSequencePlan(params: {
    prompt: string;
    parsed: {
      genre: string;
      mood: string[];
      tone: string;
      characters: Array<{ name: string; role: string; description: string }>;
      keyElements: string[];
      setting: string;
      timePeriod: string;
      location: string;
      videoType: string;
      style: string[];
      visualReferences?: string[];
    };
  }): SequencePlan {
    const { prompt, parsed } = params;
    
    // Calculate number of shots
    const targetDuration = this.config.targetDuration || 60;
    const shotsPerMinute = this.config.shotsPerMinute || 3;
    const totalShots = Math.max(3, Math.ceil((targetDuration / 60) * shotsPerMinute));
    
    // Generate sequences
    const sequences = this.generateShots(prompt, parsed, totalShots, targetDuration);
    
    // Calculate shot type distribution
    const shotTypeDistribution = this.calculateShotDistribution(sequences);
    
    return {
      planId: 'seq_' + Date.now(),
      projectTitle: parsed.genre || 'Untitled Project',
      createdAt: new Date().toISOString(),
      aspectRatio: this.config.aspectRatio || '16:9',
      totalDuration: targetDuration,
      totalShots,
      sequences,
      shotTypeDistribution,
      musicMood: this.determineMusicMood(parsed.mood, parsed.tone),
      overallMood: parsed.mood[0] || parsed.tone || 'neutral',
    };
  }

  /**
   * Generate individual shots
   */
  private generateShots(
    prompt: string,
    parsed: {
      genre: string;
      mood: string[];
      tone: string;
      characters: Array<{ name: string; role: string; description: string }>;
      keyElements: string[];
      setting: string;
      timePeriod: string;
      location: string;
      videoType: string;
      style: string[];
      visualReferences?: string[];
    },
    totalShots: number,
    duration: number
  ): SequenceShot[] {
    const shots: SequenceShot[] = [];
    
    const avgShotDuration = duration / totalShots;
    let currentTime = 0;
    
    // Determine shot pattern based on video type
    const shotPattern = this.getShotPattern(parsed.videoType);
    
    for (let shotIndex = 0; shotIndex < totalShots; shotIndex++) {
      const isFirst = shotIndex === 0;
      const isLast = shotIndex === totalShots - 1;
      const position = shotIndex / totalShots;
      
      // Determine shot type from pattern
      const shotType = this.getShotFromPattern(shotPattern, shotIndex, totalShots);
      
      // Calculate timing
      const shotDuration = this.calculateShotDuration(avgShotDuration, shotType, position, isFirst, isLast);
      const endTime = Math.min(currentTime + shotDuration, duration);
      
      // Determine camera settings
      const cameraAngle = this.getCameraAngle(parsed.mood, position);
      const cameraMovement = this.getCameraMovement(parsed.mood, shotType, position);
      const lens = this.getLens(shotType, cameraMovement);
      
      // Generate first image prompt
      const imagePrompt = this.generateImagePrompt(
        prompt,
        parsed,
        shotType,
        cameraAngle,
        cameraMovement,
        shotIndex,
        isFirst,
        isLast
      );
      
      // Generate negative prompt
      const negativePrompt = this.generateNegativePrompt(parsed.genre);
      
      // Generate style notes
      const styleNotes = this.generateStyleNotes(parsed.style, parsed.visualReferences);
      
      // Determine transition
      const transitionIn = isFirst ? 'fade' : 'cut';
      const transitionOut = isLast ? 'fade' : 'cut';
      
      // Determine intensity
      const intensity = this.determineIntensity(position, parsed.mood, parsed.tone);
      
      // Generate audio description
      const musicDescription = this.config.includeAudio 
        ? this.generateMusicDescription(parsed.mood, intensity, position)
        : '';
      
      // Generate dialogue if applicable
      const dialogue = this.config.includeDialogue && parsed.characters.length > 0
        ? this.generateDialogue(parsed.characters, currentTime, shotDuration, shotIndex)
        : [];
      
      shots.push({
        shotId: 'shot_' + (shotIndex + 1).toString().padStart(3, '0'),
        sequenceNumber: shotIndex + 1,
        shotNumber: shotIndex + 1,
        startTime: currentTime,
        endTime: endTime,
        duration: shotDuration,
        sceneDescription: this.generateSceneDescription(parsed, shotIndex, totalShots),
        sceneIndex: Math.floor((shotIndex / totalShots) * 3),
        shotType,
        cameraAngle,
        cameraMovement,
        lens,
        firstImagePrompt: imagePrompt,
        negativePrompt,
        styleNotes,
        aspectRatio: this.config.aspectRatio || '16:9',
        resolution: this.getResolution(),
        transitionIn,
        transitionOut,
        musicDescription,
        soundEffects: this.generateSoundEffects(parsed, shotType),
        dialogue,
        mood: parsed.mood[0] || 'neutral',
        intensity,
      });
      
      currentTime = endTime;
    }
    
    return shots;
  }

  /**
   * Get shot pattern based on video type
   */
  private getShotPattern(videoType: string): string[] {
    const patterns: Record<string, string[]> = {
      'trailer': ['ELS', 'LS', 'FS', 'MCU', 'CU', 'LS', 'MCU', 'FS', 'ELS'],
      'teaser': ['ELS', 'MCU', 'CU', 'MCU', 'LS'],
      'short_film': ['ELS', 'LS', 'MCU', 'CU', 'MCU', 'FS', 'LS'],
      'music_video': ['ELS', 'LS', 'MCU', 'CU', 'MCU', 'LS', 'ELS'],
      'documentary': ['LS', 'MCU', 'LS', 'CU', 'MCU'],
      'commercial': ['ELS', 'MCU', 'CU', 'MCU'],
      'unknown': ['ELS', 'LS', 'MCU', 'CU', 'LS'],
    };
    
    return patterns[videoType] || patterns['unknown'];
  }

  /**
   * Get shot type from pattern
   */
  private getShotFromPattern(pattern: string[], index: number, total: number): ShotType {
    if (pattern.length >= total) {
      return pattern[index] as ShotType;
    }
    
    const patternIndex = index % pattern.length;
    return pattern[patternIndex] as ShotType;
  }

  /**
   * Calculate shot duration
   */
  private calculateShotDuration(
    avgDuration: number,
    shotType: ShotType,
    position: number,
    isFirst: boolean,
    isLast: boolean
  ): number {
    if (isFirst) return avgDuration * 1.5;
    if (isLast) return avgDuration * 1.3;
    
    const typeModifiers: Record<ShotType, number> = {
      'ELS': 1.4,
      'LS': 1.2,
      'FS': 1.0,
      'MCU': 0.9,
      'CU': 0.7,
      'ECU': 0.5,
    };
    
    let positionModifier = 1.0;
    if (position > 0.6 && position < 0.85) {
      positionModifier = 0.85;
    } else if (position >= 0.85) {
      positionModifier = 1.1;
    }
    
    return Math.round(avgDuration * typeModifiers[shotType] * positionModifier * 10) / 10;
  }

  /**
   * Get camera angle based on mood and position
   */
  private getCameraAngle(mood: string[], position: number): CameraAngle {
    const isDark = mood.includes('dark') || mood.includes('tense');
    const isEpic = mood.includes('epic') || mood.includes('exciting');
    
    if (isEpic && position < 0.2) {
      return 'low-angle';
    }
    
    if (isDark && position > 0.7) {
      return 'high-angle';
    }
    
    if (Math.random() > 0.7) {
      return Math.random() > 0.5 ? 'low-angle' : 'high-angle';
    }
    
    return 'eye-level';
  }

  /**
   * Get camera movement based on mood, shot type, and position
   */
  private getCameraMovement(mood: string[], shotType: ShotType, position: number): CameraMovement {
    const isAction = mood.includes('exciting') || mood.includes('tense');
    
    if (shotType === 'CU' || shotType === 'ECU') {
      return 'static';
    }
    
    if (isAction) {
      const movements: CameraMovement[] = ['tracking', 'steadicam', 'dolly-in', 'dolly-out'];
      return movements[Math.floor(Math.random() * movements.length)];
    }
    
    if (shotType === 'ELS' || shotType === 'LS') {
      return Math.random() > 0.5 ? 'pan-right' : 'pan-left';
    }
    
    return 'static';
  }

  /**
   * Get lens type based on shot type and movement
   */
  private getLens(shotType: ShotType, movement: CameraMovement): LensType {
    if (movement === 'tracking' || movement === 'steadicam') {
      return 'wide';
    }
    
    if (shotType === 'ELS' || shotType === 'LS') {
      return 'wide';
    }
    
    if (shotType === 'CU' || shotType === 'ECU') {
      return 'telephoto';
    }
    
    return 'normal';
  }

  /**
   * Generate first image prompt for ComfyUI
   */
  private generateImagePrompt(
    prompt: string,
    parsed: {
      genre: string;
      mood: string[];
      tone: string;
      characters: Array<{ name: string; role: string; description: string }>;
      keyElements: string[];
      setting: string;
      timePeriod: string;
      location: string;
      style: string[];
    },
    shotType: ShotType,
    cameraAngle: CameraAngle,
    cameraMovement: CameraMovement,
    shotIndex: number,
    isFirst: boolean,
    isLast: boolean
  ): string {
    const components: string[] = [];
    
    const shotDescriptions: Record<ShotType, string> = {
      'ELS': 'extreme long shot establishing',
      'LS': 'long shot showing',
      'FS': 'full shot of',
      'MCU': 'medium close-up of',
      'CU': 'close-up on',
      'ECU': 'extreme close-up detail of',
    };
    components.push(shotDescriptions[shotType]);
    
    if (shotType === 'ELS' || shotType === 'LS') {
      if (parsed.setting !== 'unspecified') {
        components.push('a ' + parsed.setting);
      }
      if (parsed.location !== 'unspecified') {
        components.push(parsed.location);
      }
      if (parsed.timePeriod !== 'unspecified') {
        components.push('in the ' + parsed.timePeriod);
      }
    } else if (shotType === 'MCU' || shotType === 'CU' || shotType === 'ECU') {
      if (parsed.characters.length > 0 && parsed.characters[0]?.name) {
        const character = parsed.characters[0];
        components.push(character.name.toLowerCase());
        if (character.description) {
          components.push(character.description.toLowerCase());
        }
      } else {
        components.push('a character');
      }
    }
    
    if (parsed.keyElements.length > 0 && (shotType === 'ELS' || shotType === 'LS')) {
      const relevantElements = parsed.keyElements.slice(0, 2).join(', ');
      components.push('with ' + relevantElements);
    }
    
    if (parsed.mood.length > 0) {
      const moodStr = parsed.mood.join(', ');
      components.push(', ' + moodStr + ' atmosphere');
    }
    
    if (parsed.style.length > 0) {
      const styleStr = parsed.style.join(', ');
      components.push(', ' + styleStr + ' style');
    }
    
    const cameraTerms: string[] = [];
    cameraTerms.push(cameraAngle.replace('-', ' ') + ' view');
    if (cameraMovement !== 'static') {
      cameraTerms.push(cameraMovement.replace('-', ' ') + ' camera movement');
    }
    components.push(', ' + cameraTerms.join(', '));
    
    components.push(', cinematic, 8k, highly detailed, professional cinematography');
    
    if (parsed.genre === 'cyberpunk' || parsed.genre === 'sci-fi') {
      components.push(', neon lights, futuristic technology, dystopian cityscape');
    } else if (parsed.genre === 'fantasy') {
      components.push(', magical elements, epic scale, fantasy world');
    } else if (parsed.genre === 'horror') {
      components.push(', ominous lighting, shadows, tension, dread');
    }
    
    return components.filter(c => c.trim()).join(' ');
  }

  /**
   * Generate negative prompt
   */
  private generateNegativePrompt(genre: string): string {
    const base = 'low quality, blurry, distorted, ugly, bad anatomy, extra limbs, watermark, text, signature';
    
    const genreSpecific: Record<string, string> = {
      'cyberpunk': ', daylight, bright colors, cheerful',
      'fantasy': ', modern technology, sci-fi elements, realistic',
      'horror': ', bright lighting, cheerful, happy',
      'romance': ', violence, action, horror',
    };
    
    return base + (genreSpecific[genre] || '');
  }

  /**
   * Generate style notes
   */
  private generateStyleNotes(style: string[], visualReferences?: string[]): string[] {
    const notes: string[] = [];
    
    if (style.includes('cinematic')) {
      notes.push('Use cinematic color grading');
    }
    if (style.includes('vhs')) {
      notes.push('Apply VHS/retro filter');
    }
    if (style.includes('anime')) {
      notes.push('Maintain anime art style consistency');
    }
    
    if (visualReferences) {
      visualReferences.forEach(ref => {
        notes.push('Reference style from: ' + ref);
      });
    }
    
    return notes;
  }

  /**
   * Generate scene description
   */
  private generateSceneDescription(
    parsed: {
      genre: string;
      setting: string;
      timePeriod: string;
      location: string;
    },
    shotIndex: number,
    totalShots: number
  ): string {
    const position = shotIndex / totalShots;
    
    if (position < 0.2) {
      return 'Opening establishing shot - ' + (parsed.setting !== 'unspecified' ? parsed.setting : 'unknown setting') + 
        ', ' + (parsed.timePeriod !== 'unspecified' ? parsed.timePeriod : 'unknown time period') + 
        '. Sets the tone for the ' + parsed.genre + ' project.';
    }
    
    if (position < 0.8) {
      return 'Development sequence - ' + (parsed.location !== 'unspecified' ? parsed.location : 'scene location') + 
        '. Building tension and narrative progression.';
    }
    
    return 'Climactic sequence - Peak moment of the ' + parsed.genre + ' story. Maximum impact and emotional payoff.';
  }

  /**
   * Determine intensity level
   */
  private determineIntensity(
    position: number,
    mood: string[],
    tone: string
  ): 'low' | 'medium' | 'high' {
    if (position > 0.7) return 'high';
    if (position > 0.4) return 'medium';
    
    if (mood.includes('tense') || mood.includes('exciting') || tone === 'intense') {
      return 'high';
    }
    if (mood.includes('peaceful') || tone === 'calm') {
      return 'low';
    }
    
    return 'medium';
  }

  /**
   * Generate music description
   */
  private generateMusicDescription(mood: string[], intensity: 'low' | 'medium' | 'high', position: number): string {
    const descriptions: string[] = [];
    
    if (mood.includes('dark') || mood.includes('tense')) {
      descriptions.push('dark ambient drones');
    }
    if (mood.includes('epic')) {
      descriptions.push('orchestral strings, heroic brass');
    }
    if (mood.includes('melancholic')) {
      descriptions.push('piano, melancholic melody');
    }
    if (mood.includes('exciting')) {
      descriptions.push('electronic beats, high energy');
    }
    
    if (intensity === 'high') {
      descriptions.push('building crescendo');
    }
    if (intensity === 'low') {
      descriptions.push('soft, minimal arrangement');
    }
    
    if (position > 0.8) {
      descriptions.push('climactic peak');
    }
    if (position < 0.1) {
      descriptions.push('intro theme');
    }
    
    return descriptions.join(' with ');
  }

  /**
   * Generate sound effects
   */
  private generateSoundEffects(
    parsed: {
      genre: string;
      keyElements: string[];
    },
    shotType: ShotType
  ): string[] {
    const effects: string[] = [];
    
    if (parsed.genre === 'cyberpunk' || parsed.genre === 'sci-fi') {
      effects.push('city ambience', 'distant machinery', 'neon hum');
    }
    if (parsed.genre === 'fantasy') {
      effects.push('wind through trees', 'distant bells', 'magical shimmer');
    }
    if (parsed.genre === 'horror') {
      effects.push('creepy atmosphere', 'distant footsteps', 'unsettling silence');
    }
    
    parsed.keyElements.forEach(element => {
      const elemLower = element.toLowerCase();
      if (elemLower.includes('drone')) {
        effects.push('drone whir');
      }
      if (elemLower.includes('pursuit')) {
        effects.push('footsteps running', 'car engine');
      }
    });
    
    if (shotType === 'ELS') {
      effects.push('ambient environment sound');
    }
    
    return effects.slice(0, 3);
  }

  /**
   * Generate dialogue
   */
  private generateDialogue(
    characters: Array<{ name: string; role: string; description: string }>,
    startTime: number,
    duration: number,
    shotIndex: number
  ): DialogueLine[] {
    // Ensure characters array is not empty and first character has a valid name
    if (!characters || characters.length === 0 || !characters[0]?.name) {
      return [];
    }
    
    const character = characters[0];
    const dialogues = [
      'Voice-over narration by ' + character.name,
      character.name + ' speaks intently',
      character.name + ' delivers key line',
    ];
    
    return [{
      speaker: character.name,
      text: dialogues[Math.floor(Math.random() * dialogues.length)],
      startTime: startTime,
      endTime: startTime + duration,
    }];
  }

  /**
   * Get resolution based on aspect ratio
   */
  private getResolution(): string {
    const resolutions: Record<AspectRatio, string> = {
      '16:9': '1920x1080',
      '9:16': '1080x1920',
      '1:1': '1080x1080',
      '4:3': '1440x1080',
      '21:9': '2560x1080',
      'auto': '1920x1080',
    };
    
    return resolutions[this.config.aspectRatio || '16:9'];
  }

  /**
   * Calculate shot type distribution
   */
  private calculateShotDistribution(shots: SequenceShot[]): Record<ShotType, number> {
    const distribution: Record<ShotType, number> = {
      'ELS': 0,
      'LS': 0,
      'FS': 0,
      'MCU': 0,
      'CU': 0,
      'ECU': 0,
    };
    
    shots.forEach(shot => {
      distribution[shot.shotType]++;
    });
    
    return distribution;
  }

  /**
   * Determine music mood
   */
  private determineMusicMood(mood: string[], tone: string): string {
    if (mood.includes('dark') || tone === 'intense') {
      return 'tense and ominous';
    }
    if (mood.includes('epic')) {
      return 'heroic and grand';
    }
    if (mood.includes('peaceful')) {
      return 'calm and ambient';
    }
    if (mood.includes('exciting')) {
      return 'energetic and dynamic';
    }
    return 'neutral atmospheric';
  }

  /**
   * Update configuration
   */
  updateConfig(newConfig: Partial<SequencePlannerConfig>): void {
    this.config = { ...this.config, ...newConfig };
  }

  /**
   * Get current configuration
   */
  getConfig(): SequencePlannerConfig {
    return { ...this.config };
  }
}

export const sequencePlanner = new SequencePlanner();

